### PR TITLE
Get BPartners from shopware

### PR DESCRIPTION
**Pull request by cif-ps**
Upgraded shopware bpartner route to consume MF API V2

refs: https://github.com/metasfresh/metasfresh/issues/10909

### CODE REVIEW

**Review by metas-ts — CHANGES_REQUESTED** [review]
please check my comments and requests

File: backend/de.metas.adempiere.adempiere/migration/src/main/sql/postgresql/system/10-de.metas.adempiere/5584270_sys_gh10909_Add_shopware_identifier_json_path.sql [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program. If not, see
+ * <http://www.gnu.org/licenses/gpl-2.0.html>.
+ * #L%
+ */
+
```
[the remark this answers is not in the document]
Comment (cif-ps) [reply backend/de.metas.adempiere.adempiere/migration/src/main/sql/postgresql/system/10-de.metas.adempiere/5584270_sys_gh10909_Add_shopware_identifier_json_path.sql:272]: fixed

File: misc/services/camel/de-metas-camel-externalsystems/de-metas-camel-shopware6/src/main/java/de/metas/camel/externalsystems/shopware6/api/model/order/JsonOrderCustomer.java [hunk 2]
```diff
@@ -30,6 +30,8 @@
 import lombok.NonNull;
 import lombok.Value;
 
+import javax.annotation.Nullable;
+
```
[the remark this answers is not in the document]
Comment (cif-ps) [reply misc/services/camel/de-metas-camel-externalsystems/de-metas-camel-shopware6/src/main/java/de/metas/camel/externalsystems/shopware6/api/model/order/JsonOrderCustomer.java:34]: Fixed, but just FYI regarding this, I noticed that sometimes we receive null in the customerId from the order.

File: misc/services/camel/de-metas-camel-externalsystems/de-metas-camel-shopware6/src/main/java/de/metas/camel/externalsystems/shopware6/processor/BPartnerUpsertRequestProducer.java [hunk 3]
```diff
@@ -232,53 +191,41 @@ private JsonRequestLocationUpsert getUpsertLocationsRequest()
...
 
-		final JsonExternalReferenceItem externalReferenceItem = JsonExternalReferenceItem.builder()
-				.lookupItem(jsonExternalReferenceLookupItem)
-				.version(externalReferenceVersion)
-				.build();
+		if (!customIdPresent)
+		{
+			if (isBillingAddress)
+			{
+				System.out.println(orderAddressWithCustomId);
```
[the remark this answers is not in the document]
Comment (cif-ps) [reply misc/services/camel/de-metas-camel-externalsystems/de-metas-camel-shopware6/src/main/java/de/metas/camel/externalsystems/shopware6/processor/BPartnerUpsertRequestProducer.java:215]: fixed

File: misc/services/camel/de-metas-camel-externalsystems/de-metas-camel-shopware6/src/test/resources/de/metas/camel/externalsystems/shopware6/CamelUpsertBPartnerCompositeRequest.json [hunk 4]
```diff
@@ -51,10 +49,51 @@
           "locations": {
             "requestItems": [
               {
-                "locationIdentifier": "102",
+                "locationIdentifier": "ext-Shopware6-shippingAddressId_custom",
+                "location": {
+                  "active": null,
+                  "activeSet": false,
```
[the remark this answers is not in the document]
Comment (cif-ps) [reply misc/services/camel/de-metas-camel-externalsystems/de-metas-camel-shopware6/src/test/resources/de/metas/camel/externalsystems/shopware6/CamelUpsertBPartnerCompositeRequest.json:55]: fixed

File: misc/de-metas-common/de-metas-common-bpartner/src/main/java/de/metas/common/bpartner/v2/request/JsonRequestLocationUpsertItem.java [hunk 5]
```diff
@@ -44,24 +41,18 @@
 			value = LOCATION_IDENTIFIER_DOC
 					+ "If a new location is created and the request's location has no different identifier, then this identifier is stored within the newly created lcoation.") //
```
[the remark this answers is not in the document]
Comment (cif-ps) [reply misc/de-metas-common/de-metas-common-bpartner/src/main/java/de/metas/common/bpartner/v2/request/JsonRequestLocationUpsertItem.java:42]: fixed

File: backend/de.metas.business/src/main/java/de/metas/bpartner/composite/BPartner.java [hunk 6]
```diff
@@ -120,7 +120,7 @@
 	 */
 	private final String shipmentAllocationBestBeforePolicy;
 
-
+	private Boolean isIdentifiedByExternalReference;
```
[the remark this answers is not in the document]
Comment (cif-ps) [reply backend/de.metas.business/src/main/java/de/metas/bpartner/composite/BPartner.java:123]: fixed

File: backend/de.metas.business/src/main/java/de/metas/bpartner/composite/BPartnerComposite.java [hunk 7]
```diff
@@ -231,7 +231,8 @@ public BPartnerComposite deepCopy()
 	{
 		final ImmutableList.Builder<ITranslatableString> result = ImmutableList.builder();
 
-		final boolean hasLookupKey = bpartner.getId() != null
+		final boolean hasLookupKey = bpartner.getIsIdentifiedByExternalReference()
```
[the remark this answers is not in the document]
Comment (cif-ps) [reply backend/de.metas.business/src/main/java/de/metas/bpartner/composite/BPartnerComposite.java:234]: It's referring to de.metas.common.bpartner.v2.request.JsonRequestBPartnerUpsertItem#bpartnerIdentifier as you mentioned. I removed JsonRequestComposite#bPartnerReferenceCreateRequest in v2.

File: backend/de.metas.business.rest-api-impl/src/test/resources/de/metas/rest_api/v2/bpartner/bpartnercomposite/JsonRetrieverServiceTest.snap [hunk 8]
```diff
@@ -1,410 +0,0 @@
-de.metas.rest_api.bpartner.impl.v2.bpartnercomposite.JsonRetrieverServiceTest.retrieveBPartnerComposite_byBPartnerCompositeLookupKeys=[
```
[the remark this answers is not in the document]
Comment (cif-ps) [reply backend/de.metas.business.rest-api-impl/src/test/resources/de/metas/rest_api/v2/bpartner/bpartnercomposite/JsonRetrieverServiceTest.snap:1]: By mistake, added it back.

File: backend/de.metas.business/src/main/java/de/metas/bpartner/composite/BPartner.java [hunk 9]
```diff
@@ -120,7 +120,11 @@
 	 */
 	private final String shipmentAllocationBestBeforePolicy;
 
-
+	/**
+	 * If true, it means that the BPartner is valid without having a code, metasfresh or gln.
+	 * It's identifier is an external reference that is included in the BPartnerComposite instance which holds this BPartner.
```
Comment (metas-ts) [inline backend/de.metas.business/src/main/java/de/metas/bpartner/composite/BPartner.java:125]:
```suggestion
	 * It's identifier is an external reference that is provided outside of this bpartner's composite.
```

**Review by metas-ts — APPROVED** [review]

**Pull request merged**